### PR TITLE
User POST - AddUser

### LINKED ISSUE #302: Given a user, when the user POST endpoint is called, then the new user should be added to the database.
#272

### LINKED ISSUE #378: Research and add password hashing
#272 

Using only SHA to hash passwords is considered insecure and should not be done. This task will involve researching a more secure password hashing algorithm, and installing a well-known package that can easily handle password hashing for us. This will also involve implementing this when implementing the AddUser endpoint, since passwords will be hashed on the backend.

Research:

The BCrypt.Net package was chosen. This package uses the bcrypt algorithm, and it is generally accepted as a secure password hashing algorithm. This package was chosen because it is widely used, and it makes password hashing extremely easy as shown in the documentation here: https://github.com/BcryptNet/bcrypt.net

Using statement: 
`using BCryptNet = BCrypt.Net.BCrypt;`

To hash passwords:
`string passwordHash = BCryptNet.HashPassword(data.Password);`

To check a plain text password against a hashed password:
`bool passwordCorrect = BCrypt.Verify("plaintextPassword1!", passwordHash);`

Note on EnhancedHashPassword and EnhancedVerify methods in BCrypt.Net:

A conscious decision was made to not use the "Enhanced" versions of these methods. According to this article below, it is insecure to pre-hash passwords in the manner that EnhancedHashPassword does.

> An alternative approach is to pre-hash the user-supplied password with a fast algorithm such as SHA-256, and then to hash the resulting hash with bcrypt (i.e., bcrypt(base64(hmac-sha256(data:$password, key:$pepper)), $salt, $cost)). This is a dangerous (but common) practice that should be avoided due to [password shucking](https://www.youtube.com/watch?v=OQD3qDYMyYQ) and other issues when [combining bcrypt with other hash functions](https://blog.ircmaxell.com/2015/03/security-issue-combining-bcrypt-with.html).

From: https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html

### PR DESCRIPTION
**Pull request by roguedeleter**
Closes #302 
Closes #378

This PR implements the AddUser endpoint. It also implements a password hashing package called BCrypt.Net. This uses the bcrypt algorithm, and it is generally accepted as a secure password hashing algorithm. This package was chosen because it is widely used, and it makes password hashing extremely easy as shown in the documentation here: https://github.com/BcryptNet/bcrypt.net

Using statement: 
`using BCryptNet = BCrypt.Net.BCrypt;`

To hash passwords:
`string passwordHash = BCryptNet.HashPassword(data.Password);`

To check a plain text password against a hashed password:
`bool passwordCorrect = BCrypt.Verify("plaintextPassword1!", passwordHash);`

Note on EnhancedHashPassword and EnhancedVerify methods in BCrypt.Net:

A conscious decision was made to not use the "Enhanced" versions of these methods. According to this article below, it is insecure to pre-hash passwords in the manner that EnhancedHashPassword does.

> An alternative approach is to pre-hash the user-supplied password with a fast algorithm such as SHA-256, and then to hash the resulting hash with bcrypt (i.e., bcrypt(base64(hmac-sha256(data:$password, key:$pepper)), $salt, $cost)). This is a dangerous (but common) practice that should be avoided due to [password shucking](https://www.youtube.com/watch?v=OQD3qDYMyYQ) and other issues when [combining bcrypt with other hash functions](https://blog.ircmaxell.com/2015/03/security-issue-combining-bcrypt-with.html).

From: https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html

### CODE REVIEW

File: BEIMA.Backend.FT/TestObjects.cs [hunk 1]
```diff
@@ -143,6 +143,27 @@ public class Building
             public LastModified? LastModified { get; set; }
         }
 
+        public class User
```
Comment (josephmoore99) [inline BEIMA.Backend.FT/TestObjects.cs:146]: I think we also need a LastModified property here.

File: BEIMA.Backend/UserFunctions/AddUser.cs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+                                        passwordHash,
+                                        data.FirstName,
+                                        data.LastName,
+                                        data.Role);
+            }
+            catch (Exception)
+            {
+                return new BadRequestObjectResult(Resources.CouldNotParseBody);
+            }
+            // TODO: Use actual user.
```
Comment (AshlynAdamson) [inline BEIMA.Backend/UserFunctions/AddUser.cs:53]: Will this TODO be done in another user task?

File: BEIMA.Backend.FT/TestObjects.cs (hunk 1 above)
Comment (roguedeleter) [reply BEIMA.Backend.FT/TestObjects.cs:146]: Fixed 57c5fac

File: BEIMA.Backend/UserFunctions/AddUser.cs (hunk 2 above)
Comment (roguedeleter) [reply BEIMA.Backend/UserFunctions/AddUser.cs:53]: This TODO is part of all of our endpoints as of now. It is dependent on Kenny's work of adding authentication to endpoints. Once we have that, we will be able to insert the actual username of the person who is currently using the endpoint.

**Comment by roguedeleter** [comment]
All open issues have been resolved

**Review by josephmoore99 — APPROVED** [review]

**Review by AshlynAdamson — APPROVED** [review]

**Review by Kenny-Miller — APPROVED** [review]

**Pull request merged**